qiskit/optimization/algorithms/warm_start_qaoa_optimizer.py

```diff
@@ -0,0 +1,318 @@
+
+    # def get_compatibility_msg(self, problem: QuadraticProgram) -> Optional[str]:
+    #     """Checks whether a given problem can be solved with this optimizer.
+    #
+    #     Checks whether the given problem is compatible, i.e., whether the problem can be converted
+    #     to a QUBO, and otherwise, returns a message explaining the incompatibility.
+    #
+    #     Args:
+    #         problem: The optimization problem to check compatibility.
+    #
+    #     Returns:
+    #         A message describing the incompatibility.
+    #     """
+    #     return super().get_compatibility_msg(problem)
+
+    def solve(self, problem: QuadraticProgram) -> MinimumEigenOptimizationResult:
+        """Tries to solves the given problem using the optimizer.
+
+        The pre-solver is run to warm-start the solver. Next, the optimizer is run
+        to try to solve the optimization problem.
+
+        Args:
+            problem: The problem to be solved.
+
+        Returns:
+            The result of the optimizer applied to the problem.
+
+        Raises:
+            QiskitOptimizationError: If problem not compatible.
+        """
+
+        msg = self.get_compatibility_msg(problem)
+        if len(msg) > 0:
+            raise QiskitOptimizationError('Incompatible problem: {}'.format(msg))
+
+        # convert problem to QUBO or another form if converters are specified
+        qubo_problem = self._convert(problem, self._converters)
+
+        # if the pre-solver can't solve the problem then it should be relaxed.
+        if self._relax_for_pre_solver:
+            pre_solver_problem = self._relax_problem(qubo_problem)
+        else:
+            pre_solver_problem = qubo_problem
+
+        opt_result = self._pre_solver.solve(pre_solver_problem)
+        num_initial_solutions = min(self._num_initial_solutions, len(opt_result.all_solutions))
+
+        # construct operator and offset
+        operator, offset = qubo_problem.to_ising()
+
+        # todo: accessing all solutions
+        initial_solutions = opt_result.all_solutions[:num_initial_solutions]
+
+        results = []  # type: List[MinimumEigenOptimizationResult]
+        for initial_solution, _ in initial_solutions:
+            # Set the solver using the result of the pre-solver.
+            initial_variables = self._create_initial_variables(initial_solution)
+            self._qaoa.initial_state = self._create_initial_state(initial_variables)
+            self._qaoa.mixer = self._mixer_factory.create_mixer(initial_variables)
+
+            # approximate ground state of operator using min eigen solver.
+            results.append(self._solve_internal(operator, offset, qubo_problem, problem))
+
+        # todo: handle converters and _interpret
+        if len(results) == 1:
+            return results[0]
+        else:
+            return self._aggregator.aggregate(results, qubo_problem, None)
+
+    @staticmethod
+    def _relax_problem(problem: QuadraticProgram) -> QuadraticProgram:
```

Comment:
An alternative would be to have this in `QuadraticProgram` class.

Reply:
I'd keep it here, `QuadraticProgram` is really heavy. Once we see another case for relaxation, it can be moved to `QuadraticProgram`.